tez-api/src/main/java/org/apache/tez/frameworkplugins/FrameworkUtils.java

```diff
@@ -0,0 +1,96 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.tez.frameworkplugins;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+import javax.annotation.Nullable;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.tez.common.ReflectionUtils;
+import org.apache.tez.dag.api.TezConfiguration;
+import org.apache.tez.dag.api.TezConstants;
+import org.apache.tez.dag.api.TezReflectionException;
+
+public final class FrameworkUtils {
+
+  private FrameworkUtils() {}
+
+  /*
+    Searches for a FrameworkService provider which implements a target interface.
+    The interface should be either ClientFrameworkService or ServerFrameworkService.
+    Depending on which interface is used, either the client or server class of a
+    matching FrameworkMode will be used as the implementation.
+
+    NOTE: Layering of FrameworkServices in a decorator-style is currently not supported
+
+    An implementation is searched in the following order:
+     1. If conf is not null and the parameter TEZ_FRAMEWORK_MODE is set:
+       the value of TEZ_FRAMEWORK_MODE from the conf will be used
+     2. If conf is null or the parameter TEZ_FRAMEWORK_MODE is not set
+        and the environment var TEZ_FRAMEWORK_MODE is not empty:
+            the value of the environment var will be used
+     3. Otherwise:
+       the default java.util.ServiceLoader behavior will be used,
+       i.e. the implementation classname should appear in a file on the classpath at the location
+            META-INF/services/org.apache.tez.frameworkplugins.ClientFrameworkService
+            or META-INF/services/org.apache.tez.frameworkplugins.ServerFrameworkService
+   */
+  public static <T extends FrameworkService> T get(Class<T> interfaze, @Nullable Configuration conf) {
+    try {
+      if ((conf != null) && (conf.get(TezConfiguration.TEZ_FRAMEWORK_MODE) != null)) {
+        return getByMode(interfaze, conf.get(TezConfiguration.TEZ_FRAMEWORK_MODE));
+      } else if (System.getenv(TezConstants.TEZ_FRAMEWORK_MODE) != null) {
+        return getByMode(interfaze, System.getenv(TezConstants.TEZ_FRAMEWORK_MODE));
+      } else {
+        return getByServiceLoader(interfaze);
+      }
+    } catch (TezReflectionException e) {
+      throw new RuntimeException("Failed to load framework service for interface: " + interfaze.getName(), e);
+    }
+  }
+
+  private static <T extends FrameworkService> T getByServiceLoader(Class<T> interfaze) {
+    List<T> services = new ArrayList<>();
+    ServiceLoader<T> frameworkService = ServiceLoader.load(interfaze);
+    for (T service : frameworkService) {
+      services.add(service);
+    }
+    if (services.isEmpty()) {
+      return null;
```

Comment:
in the last callers we are doing `null` checks and then doing `new YarnServerFrameworkService()``, why don't we just return that here itself and save those redundant `null` checks